### ISSUE TITLE
Unresponsive quick settings tile

### ISSUE THREAD

**Issue by AdriandeLima** [beemdevelopment/Aegis#1501]
### Version

3.2

### Source

Google Play

### Vault encryption

Yes (with biometric unlock)

### Device

Pixel 6

### Android version

14

### ROM

Default

### Steps to reproduce

Add "open vault" to quick settings. 
Tap it
Wait.... Wait some more.... Wait a bit more
Oh finally the vault has launched

### What do you expect to happen?

The vault (aegis) to launch as soon as the tile is pressed

### What happens instead?

There's a significant delay between pressing the tile and anything happening, somewhere in the region of 2-5 seconds. 
Launching from the home screen icon is instant. 
I've noticed other third party tiles have similar delays so this issue might not be isolated to just Aegis. System tiles (eg internet) are fully responsive and activate instantly.

**Comment by michaelschattgen** [comment]
Unfortunately I can't reproduce this on my Pixel 9 Pro running Android 14. I will try later today on my Pixel 7 Pro XL.

Do you experience the same issues on your device? @alexbakker

**Comment by alexbakker** [comment]
I'm able to reproduce this on a Pixel 8, but only after pressing the tile right after killing Aegis. If I wait for a couple of seconds after killing Aegis before pressing the tile, the app opens quickly.

But given the fact that you also experience this with other apps, [and other users on the internet reporting the same](https://www.google.com/search?q=android+quick+settings+delay), I don't think this is an issue in Aegis.

**Comment by AdriandeLima** [comment]
What's funny, most of those other posts online are mine, it's been a long standing issue on my device that no seems able to help fix

**Comment by Kianda** [comment]
_Device: Pixel 9 Pro (always latest updates available)_

I've been hitting this issue for a while and eventually got annoyed enough to do something about it.
I'm a backend web developer who occasionally shenanigans in devops & cloud architecture (Android development is very much not my natural habitat), so please roast my changes before I embarrass myself with a proper PR.

I've been running a patched build for about a week now and the tile is **way** more responsive. The difference is very noticeable.

A debug APK (installs alongside the main app) is available in the [releases of my fork](https://github.com/Kianda/Aegis/releases) if anyone wants to try it directly. Feel free to use the pre-built APK or build your own from the [source](https://github.com/Kianda/Aegis).

---

## Root cause

The fix came from comparing Aegis's tile implementation side-by-side with Bitwarden's, which does not have this lag. Two things needed to be changed together, neither alone was enough.

### 1. `ACTIVE_TILE=true` in the manifest

Both `LaunchAppTileService` and `LaunchScannerTileService` declare:

```xml
<meta-data android:name="android.service.quicksettings.ACTIVE_TILE" android:value="true" />
```

With `ACTIVE_TILE=true`, Android does **not** bind the `TileService` when the quick settings panel opens. The service is only bound on-demand, at the moment the user taps the tile. This means the entire sequence (cold process start, Application init, service binding, `onStartListening()`, `onClick()`) happens **after** the tap, while the user is waiting.

Without `ACTIVE_TILE`, Android binds the `TileService` as soon as the quick settings panel becomes visible. The process starts in the background while the user is browsing the panel. By the time they tap, the process is already warm and `onClick()` fires immediately. This is visible as a brief disabled->enabled animation on the tile when the panel opens, identical to Bitwarden's behaviour.

Google's documentation recommends `ACTIVE_TILE=true` as more efficient for the system (less binding churn), but this optimises for battery/memory at the direct cost of perceived tap latency, a trade-off the documentation does not spell out. Bitwarden omits the metadata entirely.

### 2. `onStartListening()` calling `tile.updateTile()`

The original `onStartListening()` override:

```java
Tile tile = getQsTile();
if (tile != null) {
    tile.setState(Tile.STATE_INACTIVE);
    tile.updateTile();
}
```

`tile.updateTile()` is a Binder IPC call to SystemUI. With `ACTIVE_TILE=true`, this call happened on the tap path (since `onStartListening()` was invoked immediately before `onClick()`), adding to the delay. It is also unnecessary for a launch-only tile that carries no dynamic state to display, so the entire override was removed.

### 3. Additional cleanup

The following were also corrected to match Bitwarden's implementation:

- `Intent.FLAG_ACTIVITY_CLEAR_TASK` removed. Combined with `FLAG_ACTIVITY_NEW_TASK`, it destroyed the entire activity back stack on every tile tap, forcing a full cold activity start even when the app was already in memory.

- `intent.setAction(Intent.ACTION_MAIN)` removed. Not needed for a component-targeted intent launched from a service.

- `PendingIntent.FLAG_UPDATE_CURRENT` removed. Contradictory when combined with `FLAG_IMMUTABLE` (the intent cannot be modified).

---

## Files changed

**`app/src/main/AndroidManifest.xml`**: removed the `ACTIVE_TILE` metadata block from both tile service declarations.

**`LaunchAppTileService.java` and `LaunchScannerTileService.java`**:
- Removed `onStartListening()` override entirely.
- Removed `intent.setAction(Intent.ACTION_MAIN)`.
- Removed `Intent.FLAG_ACTIVITY_CLEAR_TASK` from intent flags.
- Changed `PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE` to `PendingIntent.FLAG_IMMUTABLE` only.
- Removed unused `android.service.quicksettings.Tile` import.

My commit: https://github.com/Kianda/Aegis/commit/d7d5e5834782ade4fa75be21c3cc5a8afc7c2c2a

**Comment by onkel-albert** [comment]
@Kianda this sounds awesome! Have you considered making a proposal (as mentioned in [CONTRIBUTING.md](https://github.com/beemdevelopment/Aegis/blob/master/CONTRIBUTING.md#pull-requests))

**Comment by Kianda** [comment]
Update: **I'm withdrawing the patch, no proposal.**

On a Pixel 10 Pro with current builds the lag is back, on the patched build and on every app with a launch tile (Amazon, Bitwarden, Aegis). Bitwarden was my control case, so the app-side explanation doesn't survive.

It also isn't a bug anyone will fix: Google's [TileService guidance](https://developer.android.com/develop/ui/views/quicksettings-tiles#decide-when-to-create) explicitly says not to use tiles to launch an app, and to use an app shortcut or launcher instead.

What I'm currently using: a shortcut in the notification shade with [Notification Shortcuts](https://play.google.com/store/apps/details?id=rk.android.app.notificationshortcuts).

The debug APK on my fork is stale, don't treat it as a fix.
